### ISSUE TITLE
Training on dataset where some images are all BG class

### ISSUE THREAD

**Issue by rbavery** [matterport/Mask_RCNN#757]
Hello, I'm trying to apply mrcnn to satellite imagery. I'm inspecting the data for errors since I was getting poor training results (I've already made adjustments for the fact that my inputs have 8 channels instead of 3.) When I use the inspect data notebook on my dataset, I see that a lot of images have no objects (over 100 out of 469). But in the nucleus example, all training images have at least one object. I think an issue I'm having is that I am not correctly encoding the mask for negative samples with no objects.

Here is my data inspection notebook with example images: https://github.com/rbavery/CropMask_RCNN/blob/master/notebooks/crops/inspect_crop_data.ipynb

For each image that has no field, I have an empty mask of shape (262, 262, 1) that is all zeros, same shape as the image except for the channels. This would be different from a mask with one field which would have 1s for fields and 0s for no field. 
`mrcnn.visualize.display_instances()` works on images with at least 1 field but fails for images with no fields. the reason is that `load_image_gt()` in mrcnn/model.py changes the shape of the masks with no objects from (262, 262, 1) to (262, 262, 0) in this step:
```
    # Note that some boxes might be all zeros if the corresponding mask got cropped out.
    # and here is to filter them out
    _idx = np.sum(mask, axis=(0, 1)) > 0
    print(_idx, 'idx')
    mask = mask[:, :, _idx]
    print(mask.shape, '1275')
    class_ids = class_ids[_idx]
    # Bounding boxes. Note that some boxes might be all zeros
    # if the corresponding mask got cropped out.
    # bbox: [num_instances, (y1, x1, y2, x2)]
    bbox = utils.extract_bboxes(mask)
``` 
This then makes mask an empty array, with no zeros to indicate the background class, causing display to fail and possibly the model training to fail (still not sure if this is what causes training issues but I think it is likely)

Any suggestions on how to encode sample masks with only the background class, no objects?

**Comment by mekomlusa** [comment]
+1 on this. I'm working on a dataset that contains no object at all in some sample images, as they're negative examples. Your thread inspires me to consider if this package has poor support on background-only images.

**Comment by rbavery** [comment]
Yeah @mekomlusa  I was wondering if discarding is the best way to go. If I tried that I'd lose about 100 out of ~500 samples, but interested to hear if that works out for you if you try it. 

Since you're not encountering NaNs in training, I think there might be another reason my rpn bbox loss decreases but eventually turns to NaN. I used the inspect model notebook to visualize what was going on and it looks like right before rpn bbox goes to nan, there are many regions proposed that are way off from the actively farmed fields in the top corners. But I'm not getting much of an idea from the inspection why rpn bbox loss goes to nan. Any suggestions are much appreciated! 

The next thing I'll try is to just train on 3 channel RGB rather than 8 channels so that I can pre-train on ImageNet instead of training from scratch. I'll also use larger inputs than ~256x256, I'm currently gridding up 8 ~2500x2500 image and can adjust the inputs to 512x512 since I'm noticing a lot of fields are partially present.

Here are my results for one field at the 20th epoch. Fails at the 21st epoch.
![region_proposals_classified](https://user-images.githubusercontent.com/22258697/42428691-fa3acedc-8302-11e8-8d70-1fe50672c1cb.png)

![selection_001](https://user-images.githubusercontent.com/22258697/42427143-ee27f73a-82fa-11e8-83f2-a57a0ee4770d.png)

Configurations:

```
Configurations:
BACKBONE                       resnet50
BACKBONE_STRIDES               [4, 8, 16, 32, 64]
BATCH_SIZE                     1
BBOX_STD_DEV                   [0.1 0.1 0.2 0.2]
CHANNELS_NUM                   8
COMPUTE_BACKBONE_SHAPE         None
DETECTION_MAX_INSTANCES        100
DETECTION_MIN_CONFIDENCE       0.7
DETECTION_NMS_THRESHOLD        0.3
FPN_CLASSIF_FC_LAYERS_SIZE     1024
GPU_COUNT                      1
GRADIENT_CLIP_NORM             5.0
IMAGES_PER_GPU                 1
IMAGE_MAX_DIM                  256
IMAGE_META_SIZE                14
IMAGE_MIN_DIM                  256
IMAGE_MIN_SCALE                0
IMAGE_RESIZE_MODE              pad64
IMAGE_SHAPE                    [256 256   8]
LEARNING_MOMENTUM              0.9
LEARNING_RATE                  0.001
LOSS_WEIGHTS                   {'rpn_class_loss': 1.0, 'rpn_bbox_loss': 1.0, 'mrcnn_class_loss': 1.0, 'mrcnn_bbox_loss': 1.0, 'mrcnn_mask_loss': 1.0}
MASK_POOL_SIZE                 14
MASK_SHAPE                     [28, 28]
MAX_GT_INSTANCES               28
MEAN_PIXEL                     [259.6  347.   259.8  416.3  228.23 313.4  187.5  562.9 ]
MINI_MASK_SHAPE                (56, 56)
NAME                           wv2-gridded-no-partial
NUM_CLASSES                    2
POOL_SIZE                      7
POST_NMS_ROIS_INFERENCE        1000
POST_NMS_ROIS_TRAINING         2000
ROI_POSITIVE_RATIO             0.33
RPN_ANCHOR_RATIOS              [0.5, 1, 2]
RPN_ANCHOR_SCALES              (16, 32, 64, 128, 180)
RPN_ANCHOR_STRIDE              1
RPN_BBOX_STD_DEV               [0.1 0.1 0.2 0.2]
RPN_NMS_THRESHOLD              0.7
RPN_TRAIN_ANCHORS_PER_IMAGE    256
STEPS_PER_EPOCH                1000
TOP_DOWN_PYRAMID_SIZE          256
TRAIN_BN                       False
TRAIN_ROIS_PER_IMAGE           100
USE_MINI_MASK                  False
USE_RPN_ROIS                   True
VALIDATION_STEPS               100
WEIGHT_DECAY                   0.0001
```

**Comment by mekomlusa** [comment]
@rbavery Thanks for sharing your config. What I'm afraid is that in reality, the input images I will feed to my trained model will be either pure background or with the objects of interest. If I train the model extensively without any "noise" data, I'm afraid that the model may fail if it's indeed a pure background with no object. But I will try (if time permits :P), though that means I'll lose half of my data (~2,000 pics in total).

As for your `rpn_bbox_loss` going to NaN - did you have this on other loss metrics as well? I did one time, like after epoch 5 all my losses suddenly become NaN (both training and validation). That's because my learning rate was set too high. The default parameter is 0.001 and looks like you didn't change yours as well, but for my case I have to set it lower or the whole model diverges.

**Comment by mekomlusa** [comment]
BTW, training extensively using only images with objects make the situation worse. My model now always "sees" stuffs while there's nothing there. Even more sensitive than before.

**Comment by rbavery** [comment]
Thanks for the update and help @mekomlusa. I think that the mrcnn model.py can handle negative samples just fine then. It was only the rpn_bbox_loss that went to NaN after decreasing for 20 epochs. Once rpn_bbox_loss went to NaN, all other losses went up in magnitude but did not go to NaN (except for overall loss). I'll try a lower learning rate if I end up getting the same error with a simpler case, instead of parsing fields/no fields I'm going to try circular fields vs everything else.

**Comment by mekomlusa** [comment]
@eljirg Yes, previously my training examples were a mixture of background only images + true samples. It's better than the other set which consists of solely true samples. But to make it work (esp. if you have only 2 binary classes) you need to tweak the code a little bit (mainly the visualization/IoU part). I didn't use the ballon example so cannot comment more on that. Good luck!

**Comment by SreenivasVRao** [comment]
@mekomlusa , can you explain how to tweak the code? I see compute_overlaps_mask in utils.py

I need to compute mIOU, but this function throws errors. My mask dimensions are (H, W, 0) at times.



 `    if masks1.shape[0] == 0 or masks2.shape[0] == 0:
        return np.zeros((masks1.shape[0], masks2.shape[-1]))
    # flatten masks and compute their areas


    masks1 = np.reshape(masks1 > .5, (-1, masks1.shape[-1])).astype(np.float32)
    masks2 = np.reshape(masks2 > .5, (-1, masks2.shape[-1])).astype(np.float32)
    area1 = np.sum(masks1, axis=0)
    area2 = np.sum(masks2, axis=0)
`

**Comment by pallaviroyal** [comment]
@mekomlusa Kindly can you suggest how to load and label negative images with their empty masks ?

**Comment by SreenivasVRao** [comment]
Sorry I don't have access to the codebase anymore. But I think this will help: https://github.com/matterport/Mask_RCNN/issues/532#issuecomment-400091605

I discarded NaN values.

**Comment by pradeeprathore04** [comment]
@SreenivasVRao  Thanks a lot for your prompt reply. I am trying the suggestion in #532

**Comment by mekomlusa** [comment]
Yep, @SreenivasVRao is right. I cannot access my original codebase anymore, but I believe my comment for that issue should work.

**Comment by smudge1872** [comment]
@pallaviroyal  Did you find out how to load the masks of background images only?

**Comment by pallaviroyal** [comment]
@smudge1872 Yes. i found .  I was doing it for [tgs-salt-identification-challenge](https://www.kaggle.com/c/tgs-salt-identification-challenge). I found how to prepare dataset object in case of Mask_RCNN.I checked my backup. Apologies i missed my code. But Mask_RCNN networks implementation is not meant for BG alone too. Because they have internally ROI(Region of Interest) layer.

**Comment by smudge1872** [comment]
@pallaviroyal  @mekomlusa   So it is not possible to add background(negative) images to the training set?

**Comment by mekomlusa** [comment]
@smudge1872 I won't say it's not possible, I got the best result when using half pure background + half normal images (with the objects of interest). Code is available in the earlier thread.

**Comment by smudge1872** [comment]
Thanks @mekomlusa .  Just to verify. 1) Use the modified  "compute_overlaps_masks" function  in your branch,  and  in the load_mask function, set masks for background only images to np.empty([0, 0, 0])  and maskIds to  np.empty([0], np.int32) .  And this will enable training with background only images?

**Comment by keineahnung2345** [comment]
You can try out this PR: https://github.com/matterport/Mask_RCNN/pull/1088.

**Comment by ghost** [comment]
> Thanks @mekomlusa . Just to verify. 1) Use the modified "compute_overlaps_masks" function in your branch, and in the load_mask function, set masks for background only images to np.empty([0, 0, 0]) and maskIds to np.empty([0], np.int32) . And this will enable training with background only images?

@smudge1872 what code edit did you do to load_mask()? I just edited the compute_overlap_masks fn and it returns an MAP of 0

**Comment by ghost** [comment]
> > Thanks @mekomlusa . Just to verify. 1) Use the modified "compute_overlaps_masks" function in your branch, and in the load_mask function, set masks for background only images to np.empty([0, 0, 0]) and maskIds to np.empty([0], np.int32) . And this will enable training with background only images?
> 
> @smudge1872 what code edit did you do to load_mask()? I just edited the compute_overlap_masks fn and it returns an MAP of 0

Anyone know the answer to this?

**Comment by smudge1872** [comment]
@sohinimallick   I used the balloon.py example as a template to create a config class for the object I was trying to detect.  I edited the load_mask function in balloon.py example.  I used pixel ground truthed masks for the images that had the objects. For images that did not have the objects. I did this
`mask = np.empty([0, 0, 0]);
			maskIDs = np.empty([0], np.int32);
			return mask, maskIDs;`
